feat(data_structures): introduce `PointerExt` trait

overlookmotel opened the pull request:
Rust v1.87.0 introduced new methods for pointers, notably `offset_from_unsigned` which calculates the distance between 2 pointers.

https://doc.rust-lang.org/std/primitive.pointer.html#method.offset_from_unsigned

`offset_from_unsigned` is the same as `offset_from`, except it require the source pointer to be before the target, and so returns a `usize` rather than an `isize`. In return for this restriction, the compiler is able to generate more efficient code in some cases.

We want to use these methods, but our MSRV won't hit 1.87.0 for 3 months.

This PR introduces a `PointerExt` trait which provides similar methods `offset_from_usize` and `byte_offset_from_usize`.

On versions of Rust >= 1.87.0, these delegate directly to the native Rust methods `offset_from_unsigned` and `byte_offset_from_unsigned`. On earlier versions of Rust, they use a fallback implementation.

This PR replaces all use of `offset_from` with `offset_from_usize` in our crates.

This will probably make little difference to perf, as we don't use these methods much, and the perf advantage is small anyway. But reason I'm making this PR now is that allocator will make liberal use of them on very hot paths.

overlookmotel commented:
* **#11095** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11095?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11095?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11095).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **May 18, 3:00 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **May 18, 3:00 AM EDT**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
feat(data_structures): introduce `PointerExt` trait (#11095)

Rust v1.87.0 introduced new methods for pointers, notably `offset_from_unsigned` which calculates the distance between 2 pointers.

https://doc.rust-lang.org/std/primitive.pointer.html#method.offset_from_unsigned

`offset_from_unsigned` is the same as `offset_from`, except it require the source pointer to be before the target, and so returns a `usize` rather than an `isize`. In return for this restriction, the compiler is able to generate more efficient code in some cases.

We want to use these methods, but our MSRV won't hit 1.87.0 for 3 months.

This PR introduces a `PointerExt` trait which provides similar methods `offset_from_usize` and `byte_offset_from_usize`.

On versions of Rust >= 1.87.0, these delegate directly to the native Rust methods `offset_from_unsigned` and `byte_offset_from_unsigned`. On earlier versions of Rust, they use a fallback implementation.

This PR replaces all use of `offset_from` with `offset_from_usize` in our crates.

This will probably make little difference to perf, as we don't use these methods much, and the perf advantage is small anyway. But reason I'm making this PR now is that allocator will make liberal use of them on very hot paths.
